### PR TITLE
Fixes #PH58796 -  Enable Hostname Verification by Default and update test case

### LINKED ISSUE #29392: CVE-2023-50314 CVSS 5.3 - Enable Hostname Verification by default
### Now the configurations below have hostname verification enabled by default:
1. [SSL Default](https://openliberty.io/docs/latest/reference/config/sslDefault.html#:~:text=httpHostNameVerification) Configuration:
    * `<sslDefault httpHostNameVerification=”false” />` maintain default value to false
    * This flag enables hostname verification exclusively for java.net.HttpURLConnection connections, which exclude other SSL/TLS connections other than HttpURLConnection.
2. Individual [SSL ](https://openliberty.io/docs/latest/reference/config/ssl.html#:~:text=trusted%20certificate%20entries.-,verifyHostname,-boolean)Configuration:
    * `<ssl id=<any-ssl-alias> verifyHostname=”true” />`
    * Hostname verification is enabled for specific SSL configurations identified by their alias. This enforces SSLSocket to perform hostname verification during SSL/TLS handshake.
3. [OpenID Connect Client](https://openliberty.io/docs/latest/reference/config/openidConnectClient.html#:~:text=hostNameVerificationEnabled) Configuration: <openidConnectClient hostNameVerificationEnabled=”true” /> Hostname verification is now enabled by default for OpenID Connect clients.
4. Added Warning for when hostname verification is Disabled

### Disabling hostname verification:
1. LDAP via SSL / LDAPS is handled by the JDK and can be disabled by setting the system property `com.sun.jndi.ldap.object.disableEndpointIdentification` to true.
2. to disable hostname verification for outbound HTTPS connections, TWO flags need to be set to false: 
    1. `<sslDefault httpHostNameVerification=”false” /> AND <ssl id=”defaultSSLConfig” verifyHostname=”false” />`
    2. this is because sslDefault uses “defaultSSLConfig” as the default “sslRef”, which implicitly has verifyHostname=”true” by default. Therefore, disabling httpHostNameVerification=”false” in the <sslDefault/> alone will NOT suffice.
3. for JAX-RS and JAX-WS:
    1. “disableCNCheck” property need to be set to true to disable hostname verification. ([JAX-RS](https://openliberty.io/docs/latest/reference/config/webTarget.html#:~:text=timeout%20programmatic%20property.-,disableCNCheck,-boolean) and [JAX-WS](https://www.ibm.com/docs/en/was-liberty/core?topic=liberty-ws-bndxml-file))
    2. on top of that, “[sslConfig](https://openliberty.io/docs/latest/reference/config/webTarget.html#:~:text=timeout%20programmatic%20property.-,sslConfig,-string)”(for JAX-RS) / “[sslRef](https://www.ibm.com/docs/en/was-liberty/core?topic=liberty-ws-bndxml-file#:~:text=can%20be%20encoded.-,ssl%2Dref,-%3A%20This%20attribute%20is)”(for JAX-WS) that is used need to also set “verifyHostname=”false”” for the same reason as the point above.

### PR DESCRIPTION
**Pull request by karel-harjono**
Resolves #29392

### CODE REVIEW

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

File: dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/config/WSKeyStore.java [hunk 1]
```diff
@@ -1760,81 +1751,6 @@ private String createCertSANInfo(String hostname) {
...
-        if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled())
-            Tr.exit(tc, "getInetAddress: " + addr);
-        return addr;
-    }
-
-    /*
-     * checking for a valid dnsName value, the SAN entry is pretty specific.
-     * Only alphas, digits, and hyphin are allowed. A period is allowed when domain are added.
-     * No part of the domain name can start with a digit
-     * The dnsName can not start or end with a period, and there can not be any empty component of the domain name
```
Comment (karel-harjono) [inline dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/config/WSKeyStore.java:1785]: "there can not be any empty component of the domain name", example for empty component is "google..com"

**Comment by karel-harjono** [comment]
#

**Comment by karel-harjono** [comment]
#build

File: dev/com.ibm.ws.ssl/resources/com/ibm/ws/ssl/resources/ssl.nlsprops [hunk 2]
```diff
@@ -635,7 +635,7 @@ ssl.client.handshake.error.CWPKI0823E=CWPKI0823E: SSL HANDSHAKE FAILURE:  A sign
 ssl.client.handshake.error.CWPKI0823E.explanation=An error occurred during the SSL handshake.  It might require a signer export/import from the target host to the client TrustStore.
 ssl.client.handshake.error.CWPKI0823E.useraction=Review the extended error message from the TrustManager to determine what needs to change between the target SSL configuration and the client SSL configuration.
 # -------------------------------------------------------------------------------------------------
-ssl.client.handshake.error.CWPKI0824E=CWPKI0824E: SSL HANDSHAKE FAILURE:  Host name verification error while connecting to host [{0}].  The host name used to access the server does not match the server certificate''s SubjectDN or Subject Alternative Name information.  The extended error message from the SSL handshake exception is: [{1}].
+ssl.client.handshake.error.CWPKI0824E=CWPKI0824E: SSL HANDSHAKE FAILURE:  Host name verification error while connecting to host [{0}].  The host name used to access the server does not match the server certificate''s SubjectDN [{1}] or Subject Alternative Name information [{2}].  The extended error message from the SSL handshake exception is: [{3}].
```
Comment (karel-harjono) [inline dev/com.ibm.ws.ssl/resources/com/ibm/ws/ssl/resources/ssl.nlsprops]: clarify to display only SAN or SubjectDN

File: dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/core/WSX509TrustManager.java [hunk 3]
```diff
@@ -958,4 +960,27 @@ private void processServerTrustError(X509Certificate[] chain, String authType, S
         }
     }
 
+    private String getHostnameVerificationSANInfo(X509Certificate[] chain) {
```
Comment (karel-harjono) [inline dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/core/WSX509TrustManager.java]: update name to: "getSANInfoForHostnameVerification"

**Comment by karel-harjono** [comment]
#build

**Comment by karel-harjono** [comment]
#build

File: dev/com.ibm.ws.channel.ssl/src/com/ibm/ws/channel/ssl/internal/SSLUtils.java [hunk 4]
```diff
@@ -1268,6 +1268,16 @@ private static void configureEngine(SSLEngine engine, FlowType type, SSLLinkConf
         } else {
             // Update engine with client side specific config parameters.
             engine.setUseClientMode(true);
+            boolean verifyHostname = config.getBooleanProperty(Constants.SSLPROP_HOSTNAME_VERIFICATION);
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "SSLLinkConfig.getBooleanProperty -> " + Constants.SSLPROP_HOSTNAME_VERIFICATION + "=" + verifyHostname);
+            }
+            if (verifyHostname) {
+                sslParameters.setEndpointIdentificationAlgorithm("HTTPS");  // enable hostname verification
```
Comment (karel-harjono) [inline dev/com.ibm.ws.channel.ssl/src/com/ibm/ws/channel/ssl/internal/SSLUtils.java]: inbound flow might need HTTPS as setEndpointId Alg

File: dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/protocol/LibertySSLSocketFactory.java [hunk 5]
```diff
@@ -724,9 +723,23 @@ private static SSLParameters createSSLParameters(Properties sslprops, SSLSocket
...
 
             //Enable hostname verification
-            String enableEndpointId = sslprops.getProperty(Constants.SSLPROP_HOSTNAME_VERIFICATION);
-            if (enableEndpointId != null && enableEndpointId.equalsIgnoreCase("true")) {
-                p.setEndpointIdentificationAlgorithm(ENDPOINT_ALGORITHM);
+            String verifyHostname = sslprops.getProperty(Constants.SSLPROP_HOSTNAME_VERIFICATION);
+            if (verifyHostname != null && verifyHostname.equalsIgnoreCase("true")) {
+                if (tc.isDebugEnabled())
+                    Tr.debug(tc, "verifyHostname = " + verifyHostname);
+                if (verifyHostname != null && verifyHostname.equalsIgnoreCase("true")) {
```
Comment (karel-harjono) [inline dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/protocol/LibertySSLSocketFactory.java]: might be copied by accident here

File: dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/protocol/LibertySSLSocketFactory.java [hunk 6]
```diff
@@ -724,9 +723,23 @@ private static SSLParameters createSSLParameters(Properties sslprops, SSLSocket
...
+                    String skipHostList = sslprops.getProperty(Constants.SSLPROP_SKIP_HOSTNAME_VERIFICATION_FOR_HOSTS);
+                    if (!SSLPropertyUtils.isSkipHostnameVerificationForHosts(remoteHostname, skipHostList)) {
```
Comment (karel-harjono) [inline dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/protocol/LibertySSLSocketFactory.java]: add bracket here for clarification !

File: dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/protocol/LibertySSLSocketFactory.java [hunk 7]
```diff
@@ -813,4 +826,43 @@ public SSLSocketFactory run() throws Exception {
             }
         }
     }
+
```
Comment (karel-harjono) [inline dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/protocol/LibertySSLSocketFactory.java:812]: don't forget to remove this

File: dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/SSLPropertyUtils.java [hunk 8]
```diff
@@ -197,15 +201,67 @@ private static SSLParameters createSSLParameters(Properties properties, SSLParam
...
+     * @param String skipHostList - comma separated list of hostnames with hostname verification disabled, e.g. "hello.com, world.com"
+     */
+    public static boolean isSkipHostnameVerificationForHosts(String remoteHost, String skipHostList) {
+        if (tc.isEntryEnabled())
+            Tr.entry(tc, "isSkipHostnameVerificationForHosts", new Object[] { remoteHost, skipHostList });
+        boolean skipHostnameVerification = false;
+        if (remoteHost != null && skipHostList != null && !!!skipHostList.isEmpty()) {
+            List<String> skipHosts = Arrays.asList(skipHostList.split("\\s*,\\s*"));
+
+            for (String template : skipHosts) {
```
Comment (karel-harjono) [inline dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/SSLPropertyUtils.java]: change template to host

File: dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/SSLPropertyUtils.java [hunk 9]
```diff
@@ -197,15 +201,67 @@ private static SSLParameters createSSLParameters(Properties properties, SSLParam
...
     }
+
+    /**
+     * https://datatracker.ietf.org/doc/html/rfc2830#section-3.6
+     * The "*" wildcard character is allowed. If present, it applies only to the left-most name component.
+     *
+     * @param String host - target host
+     * @param String skipHostList - comma separated list of hostnames with hostname verification disabled, e.g. "hello.com, world.com"
+     */
+    public static boolean isSkipHostnameVerificationForHosts(String remoteHost, String skipHostList) {
```
Comment (karel-harjono) [inline dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/SSLPropertyUtils.java]: there are other place that doesn't use SSLPropertyUtils.

File: dev/com.ibm.ws.ssl/resources/com/ibm/ws/ssl/resources/ssl.nlsprops [hunk 10]
```diff
@@ -227,8 +227,9 @@ ssl.config.not.using.fips.CWPKI0049W=CWPKI0049W: UseFIPS is enabled but the SSL
 ssl.config.not.using.fips.CWPKI0049W.explanation=Although UseFIPS has been enabled, the FIPS-approved cryptographic algorithms cannot be used because the SSL configuration is not using a FIPS-approved JSSE Provider.
 ssl.config.not.using.fips.CWPKI0049W.useraction=To ensure the use of FIPS-approved cryptographic algorithms, modify the SSL configuration to use a FIPS-approved JSSE Provider.
 # -------------------------------------------------------------------------------------------------
-
-
+ssl.san.warning.CWPKI0050W=CWPKI0050W: The certificate with the {0} alias from the {1} keyStore has no subject alternative name. HostName verification is disabled for serverIdentity collective controller, replica and member SSL communication.
```
Comment (Zech-Hein) [inline dev/com.ibm.ws.ssl/resources/com/ibm/ws/ssl/resources/ssl.nlsprops]: perhaps reword "disabled" to "not enforced" for serverIdentity

File: dev/com.ibm.ws.ssl/resources/com/ibm/ws/ssl/resources/ssl.nlsprops [hunk 11]
```diff
@@ -227,8 +227,9 @@ ssl.config.not.using.fips.CWPKI0049W=CWPKI0049W: UseFIPS is enabled but the SSL
...
+ssl.san.warning.CWPKI0050W.explanation=A certificate does not have subject alternative name in the keyStore.
```
Comment (Zech-Hein) [inline dev/com.ibm.ws.ssl/resources/com/ibm/ws/ssl/resources/ssl.nlsprops]:
```suggestion
ssl.san.warning.CWPKI0050W.explanation=The default serverIdentity certificate does not have subject alternative name in the keyStore. The hostName verification check is skipped for serverIdentity certificates
```

File: dev/com.ibm.ws.ssl/resources/com/ibm/ws/ssl/resources/ssl.nlsprops [hunk 12]
```diff
@@ -227,8 +227,9 @@ ssl.config.not.using.fips.CWPKI0049W=CWPKI0049W: UseFIPS is enabled but the SSL
...
+ssl.san.warning.CWPKI0050W.useraction=Open the keyStore and check the subject alternative name on all certificates in the keyStore. Generate new certificates to enable the hostname verification.
```
Comment (Zech-Hein) [inline dev/com.ibm.ws.ssl/resources/com/ibm/ws/ssl/resources/ssl.nlsprops]:
```suggestion
ssl.san.warning.CWPKI0050W.useraction=Open the keyStore and check the subject alternative name on all certificates in the keyStore. Use the genKey and genKeyController commands to generate new certificates to enable the hostname verification.
```

File: dev/com.ibm.ws.ssl/resources/com/ibm/ws/ssl/resources/ssl.nlsprops [hunk 13]
```diff
@@ -635,7 +636,7 @@ ssl.client.handshake.error.CWPKI0823E=CWPKI0823E: SSL HANDSHAKE FAILURE:  A sign
...
 # -------------------------------------------------------------------------------------------------
-ssl.client.handshake.error.CWPKI0824E=CWPKI0824E: SSL HANDSHAKE FAILURE:  Host name verification error while connecting to host [{0}].  The host name used to access the server does not match the server certificate''s SubjectDN or Subject Alternative Name information.  The extended error message from the SSL handshake exception is: [{1}].
+ssl.client.handshake.error.CWPKI0824E=CWPKI0824E: SSL HANDSHAKE FAILURE:  Host name verification error while connecting to host [{0}].  The host name used to access the server does not match the server certificate''s [{1}].  The extended error message from the SSL handshake exception is: [{2}].
 ssl.client.handshake.error.CWPKI0824E.explanation=An error occurred during the SSL handshake.  Host name verification is enabled and server's identity can not be verified.  The host name used to access the server does not match the server certificate's SubjectDN or Subject Alternative Name information.
```
Comment (Zech-Hein) [inline dev/com.ibm.ws.ssl/resources/com/ibm/ws/ssl/resources/ssl.nlsprops]:
```suggestion
ssl.client.handshake.error.CWPKI0824E.explanation=An error occurred during the SSL handshake.  Host name verification is enabled and server's identity can not be verified.  The host name used to access the server does not match the server certificate's Subject Alternative Name information or SubjectDN .
```

File: dev/com.ibm.ws.ssl/resources/com/ibm/ws/ssl/resources/ssl.nlsprops (hunk 10 above)
Comment (karel-harjono) [reply dev/com.ibm.ws.ssl/resources/com/ibm/ws/ssl/resources/ssl.nlsprops]:
```suggestion
ssl.san.warning.CWPKI0050W=CWPKI0050W: The certificate with the {0} alias from the {1} keyStore has no subject alternative name. HostName verification is not enforced for serverIdentity collective controller, replica and member SSL communication.
```

File: dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/protocol/LibertySSLSocketFactory.java (hunk 5 above)
Comment (karel-harjono) [reply dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/protocol/LibertySSLSocketFactory.java]:
```suggestion
```

**Comment by karel-harjono** [comment]
#build

**Review by utle — APPROVED** [review]

**Review by utle — APPROVED** [review]

**Comment by clarkek123** [comment]
Fixes #PH58796

**Comment by clarkek123** [comment]
This PR addresses issue https://github.com/OpenLiberty/open-liberty/issues/29392

**Pull request merged**